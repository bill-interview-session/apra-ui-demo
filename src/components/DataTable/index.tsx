'use client'

import { useMemo, useState } from 'react'
import { Wrapper } from './styled'
import { Table, TablePaginationConfig } from 'antd'
import DataTableSearch from './search'

type Column = {
  key: string,
  title: string,
  render?: (cell: any, row?: Record<string, any>) => any
}

export type DataTableRequest = {
  query: string;
  page: number,
  limit: number,
}

type DataTableProps = {
  columns: Column[],
  data: object[],
  pagination?: TablePaginationConfig,
  search?: boolean,
  onRequest?: (request: DataTableRequest) => void,
}

const DataTable = (props: DataTableProps) => {
  const { columns: propColumns, data: propData, pagination, search, onRequest } = props || {}

  const [data, setData] = useState(propData)

  const [requestState, setRequestState] = useState<DataTableRequest>({ query: '', page: 1, limit: 10 })

  const columns = useMemo<object[]>(
    () => propColumns.map((column: Column) => ({ ...column, text: column.title, dataIndex: column.key })),
    [propColumns]
  )

  const changeRequestState = (data: Partial<DataTableRequest>) => {
    const nextRequest = { ...requestState, ...data }
    setRequestState(nextRequest)
    onRequest && onRequest(nextRequest)

  }

  return (
    <Wrapper>
      {search && (
        <DataTableSearch
          placeholder="Search..."
          onSearch={(value) => changeRequestState({ query: value })}
        />
      )}
      <Table
        dataSource={data}
        columns={columns}
        pagination={pagination}
        onChange={(pag) => changeRequestState({ page: pag.current, limit: pag.pageSize })}
      />
    </Wrapper>
  );
}

export default DataTable
