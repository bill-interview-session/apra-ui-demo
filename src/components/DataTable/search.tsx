import { Input } from "antd"
import { useState } from "react"

type DataTableSearchProps = {
  keyword?: string,
  placeholder?: string,
  onSearch: (keyword: string) => void,
}

const DataTableSearch = (props: DataTableSearchProps) => {
  const { keyword: keywordDefault, onSearch, placeholder } = props || {}

  const [keyword, setKeyword] = useState(keywordDefault ?? '')

  return (
    <Input.Search
      placeholder={placeholder ?? 'Search...'}
      value={keyword}
      onChange={(e) => setKeyword(e.target.value)}
      onSearch={onSearch}
      allowClear
      style={{ marginBottom: 16 }}
    />
  )
}

export default DataTableSearch
