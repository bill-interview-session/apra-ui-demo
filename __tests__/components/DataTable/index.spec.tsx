import { render } from '@testing-library/react';

import { TablePaginationConfig } from 'antd';

import DataTable from '@components/DataTable'

const columns = [
  {
    key: 'id',
    title: 'ID',
  },
  {
    key: 'title',
    title: 'Title',
  },
];

const data = [
  {
    id: 1,
    title: 'A',
  },
  {
    id: 2,
    title: 'B',
  }
];

describe('DataTable component', () => {
  it('should render snapshot', () => {
    const handleRequest = (_: any) => {}

    const { container } = render(
      <DataTable
        columns={columns}
        data={data ?? []}
        pagination={{
          pageSize: 10,
          current: 1,
          total: data?.length,
        } as TablePaginationConfig}
        onRequest={handleRequest}
        loading={false}
        search
      />
    );
    expect(container).toMatchSnapshot();
  });
});

