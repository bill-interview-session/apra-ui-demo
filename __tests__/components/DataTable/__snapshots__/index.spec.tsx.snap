// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable component should render snapshot 1`] = `
<div>
  <div
    class="sc-bc98d7e0-0 fYWJZw"
  >
    <span
      class="ant-input-group-wrapper ant-input-search"
      style="margin-bottom: 16px;"
    >
      <span
        class="ant-input-wrapper ant-input-group"
      >
        <span
          class="ant-input-affix-wrapper"
        >
          <input
            class="ant-input"
            placeholder="Search..."
            type="text"
            value=""
          />
          <span
            class="ant-input-suffix"
          >
            <span
              class="ant-input-clear-icon ant-input-clear-icon-hidden"
              role="button"
              tabindex="-1"
            >
              <span
                aria-label="close-circle"
                class="anticon anticon-close-circle"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="close-circle"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                  />
                </svg>
              </span>
            </span>
          </span>
        </span>
        <span
          class="ant-input-group-addon"
        >
          <button
            class="ant-btn ant-btn-default ant-btn-icon-only ant-input-search-button"
            type="button"
          >
            <span
              aria-label="search"
              class="anticon anticon-search"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="search"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                />
              </svg>
            </span>
          </button>
        </span>
      </span>
    </span>
    <div
      class="ant-table-wrapper"
    >
      <div
        class="ant-spin-nested-loading"
      >
        <div
          class="ant-spin-container"
        >
          <div
            class="ant-table"
          >
            <div
              class="ant-table-container"
            >
              <div
                class="ant-table-content"
              >
                <table
                  style="table-layout: auto;"
                >
                  <colgroup />
                  <thead
                    class="ant-table-thead"
                  >
                    <tr>
                      <th
                        class="ant-table-cell"
                      >
                        ID
                      </th>
                      <th
                        class="ant-table-cell"
                      >
                        Title
                      </th>
                    </tr>
                  </thead>
                  <tbody
                    class="ant-table-tbody"
                  >
                    <tr
                      class="ant-table-row ant-table-row-level-0"
                    >
                      <td
                        class="ant-table-cell"
                      >
                        1
                      </td>
                      <td
                        class="ant-table-cell"
                      >
                        A
                      </td>
                    </tr>
                    <tr
                      class="ant-table-row ant-table-row-level-0"
                    >
                      <td
                        class="ant-table-cell"
                      >
                        2
                      </td>
                      <td
                        class="ant-table-cell"
                      >
                        B
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
          <ul
            class="ant-pagination ant-table-pagination ant-table-pagination-right"
          >
            <li
              aria-disabled="true"
              class="ant-pagination-prev ant-pagination-disabled"
              title="Previous Page"
            >
              <button
                class="ant-pagination-item-link"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <span
                  aria-label="left"
                  class="anticon anticon-left"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="left"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
                    />
                  </svg>
                </span>
              </button>
            </li>
            <li
              class="ant-pagination-item ant-pagination-item-1 ant-pagination-item-active"
              tabindex="0"
              title="1"
            >
              <a
                rel="nofollow"
              >
                1
              </a>
            </li>
            <li
              aria-disabled="true"
              class="ant-pagination-next ant-pagination-disabled"
              title="Next Page"
            >
              <button
                class="ant-pagination-item-link"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <span
                  aria-label="right"
                  class="anticon anticon-right"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="right"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
                    />
                  </svg>
                </span>
              </button>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;
